File: lib/git.js
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

/**
 * Parse owner/repo format
 */
function parseRepo(repo) {
  const [owner, name] = repo.split('/');
  if (!owner || !name) {
    throw new Error(`Invalid repo format: ${repo}. Expected: owner/repo`);
  }
  return { owner, name };
}

/**
 * Build clone URL
 */
function buildCloneUrl(repo) {
  const { owner, name } = parseRepo(repo);

  let url = `https://github.com/${owner}/${name}.git`;

  // Proxy configuration is handled in cloneRepo function

  return url;
}

/**
 * Clone repository with proxy support
 */
function cloneRepo(repo, cacheDir, proxy = null) {
  const { owner, name } = parseRepo(repo);
  const repoPath = path.join(cacheDir, owner, name);

  // Check if already cloned
  if (fs.existsSync(repoPath)) {
    try {
      // Try to fetch latest changes
      const gitProxy = proxy ? `-c http.proxy=${proxy}` : '';
      execSync(`git ${gitProxy} -C "${repoPath}" fetch origin`, {
        stdio: 'ignore',
        timeout: 30000
      });
      execSync(`git -C "${repoPath}" reset --hard origin/HEAD`, {
        stdio: 'ignore',
        timeout: 30000
      });
      return repoPath;
    } catch (error) {
      // If fetch fails, re-clone
      fs.rmSync(repoPath, { recursive: true, force: true });
    }
  }

  // Clone repository
  try {
    fs.mkdirSync(path.dirname(repoPath), { recursive: true });

    let cloneCmd = 'git clone';
    if (proxy) {
      cloneCmd += ` -c http.proxy=${proxy}`;
    }
    cloneCmd += ' --depth 1';
    cloneCmd += ` ${buildCloneUrl(repo, proxy)}`;
    cloneCmd += ` "${repoPath}"`;

    console.log(`Cloning ${repo}...`);
    execSync(cloneCmd, {
      stdio: 'inherit',
      timeout: 120000 // 2 minutes timeout
    });

    return repoPath;
  } catch (error) {
    throw new Error(`Failed to clone ${repo}: ${error.message}`);
  }
}

/**
 * Get repository info
 */
function getRepoInfo(repoPath) {
  try {
    // Get remote URL
    const remote = execSync('git config --get remote.origin.url', {
      cwd: repoPath,
      encoding: 'utf-8'
    }).trim();

    // Get current commit
    const commit = execSync('git rev-parse HEAD', {
      cwd: repoPath,
      encoding: 'utf-8'
    }).trim();

    // Get latest commit date
    const date = execSync('git log -1 --format=%ci', {
      cwd: repoPath,
      encoding: 'utf-8'
    }).trim();

    return { remote, commit, date };
  } catch (error) {
    return null;
  }
}

module.exports = {
  parseRepo,
  buildCloneUrl,
  cloneRepo,
  getRepoInfo
};

File: internal/git/git.go
// Package git provides GitHub repository operations
package git

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

// RepoInfo holds repository metadata
type RepoInfo struct {
	Remote string `json:"remote"`
	Commit string `json:"commit"`
	Date   string `json:"date"`
}

// ParseRepo parses owner/repo format
func ParseRepo(repo string) (owner, name string, err error) {
	parts := strings.Split(repo, "/")
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return "", "", fmt.Errorf("invalid repo format: %s. Expected: owner/repo", repo)
	}
	return parts[0], parts[1], nil
}

// BuildCloneURL builds the HTTPS clone URL
func BuildCloneURL(owner, name string) string {
	return fmt.Sprintf("https://github.com/%s/%s.git", owner, name)
}

// CloneRepo clones or updates a repository
func CloneRepo(repo, cacheDir, proxy string) (string, error) {
	owner, name, err := ParseRepo(repo)
	if err != nil {
		return "", err
	}

	repoPath := filepath.Join(cacheDir, owner, name)

	// Check if already cloned
	if _, err := os.Stat(repoPath); err == nil {
		// Try to fetch latest changes
		if err := fetchLatest(repoPath, proxy); err == nil {
			return repoPath, nil
		}
		// If fetch fails, remove and re-clone
		os.RemoveAll(repoPath)
	}

	// Clone repository
	if err := os.MkdirAll(filepath.Dir(repoPath), 0755); err != nil {
		return "", fmt.Errorf("failed to create cache directory: %w", err)
	}

	url := BuildCloneURL(owner, name)
	args := []string{"clone", "--depth", "1", url, repoPath}
	args = appendGitProxyArgs(args, proxy)

	fmt.Printf("Cloning %s...\n", repo)
	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
	defer cancel()
	cmd := exec.CommandContext(ctx, "git", args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Stdin = os.Stdin

	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("failed to clone %s: %w", repo, err)
	}

	return repoPath, nil
}

// appendGitProxyArgs prepends proxy config args if proxy is set
func appendGitProxyArgs(args []string, proxy string) []string {
	if proxy == "" {
		return args
	}
	return append([]string{
		"-c", "http.proxy=" + proxy,
		"-c", "https.proxy=" + proxy,
	}, args...)
}

// fetchLatest fetches latest changes and resets to HEAD
func fetchLatest(repoPath, proxy string) error {
	gitArgs := []string{"-C", repoPath}
	gitArgs = appendGitProxyArgs(gitArgs, proxy)

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	fetchArgs := append(gitArgs, "fetch", "origin")
	if err := exec.CommandContext(ctx, "git", fetchArgs...).Run(); err != nil {
		return err
	}

	resetArgs := append(gitArgs, "reset", "--hard", "origin/HEAD")
	return exec.CommandContext(ctx, "git", resetArgs...).Run()
}

// GetRepoInfo retrieves repository metadata
func GetRepoInfo(repoPath string) *RepoInfo {
	info := &RepoInfo{}

	remote, err := exec.Command("git", "-C", repoPath, "config", "--get", "remote.origin.url").Output()
	if err != nil {
		return nil
	}
	info.Remote = strings.TrimSpace(string(remote))

	commit, err := exec.Command("git", "-C", repoPath, "rev-parse", "HEAD").Output()
	if err != nil {
		return nil
	}
	info.Commit = strings.TrimSpace(string(commit))

	date, err := exec.Command("git", "-C", repoPath, "log", "-1", "--format=%ci").Output()
	if err != nil {
		return nil
	}
	info.Date = strings.TrimSpace(string(date))

	return info
}
